src/runtime/local/datastructures/COOMatrix.h

```diff
@@ -0,0 +1,575 @@
+ * or `size_t(-1)` (indice arrays). This indicates how many entries the matrix currently holds
+ * contrary to its total capacity and needs to be set again if overwritten.
+ *
+ * Each instance of this class might represent a (row-based) sub-matrix of another
+ * `COOMatrix`. Thus, passing the `rowIdx` to `getValues()` or the `get*Idxs()`
+ * methods returns a pointer to the start of the respective column. Its size can
+ * be retrieved by calling `getNumNonZerosRow()` with the same `rowIdx`.
+ */
+template<typename ValueType>
+class COOMatrix : public Matrix<ValueType> {
+    // `using`, so that we do not need to prefix each occurrence of these
+    // fields from the super-classes.
+    using Matrix<ValueType>::numRows;
+    using Matrix<ValueType>::numCols;
+
+    /**
+     * @brief The maximum number of non-zero values this matrix was allocated
+     * to accommodate.
+     */
+    size_t maxNumNonZeros;
+
+    /**
+     * @brief Lower and upper bounds to accommodate views on a row-based
+     * sub-matrix.
+     */
+    size_t lowerRow;
+    size_t upperRow;
+
+    size_t appendHelp = 0;
+    bool view;
+
+    std::shared_ptr<ValueType> values;
+    std::shared_ptr<size_t> colIdxs;
+    std::shared_ptr<size_t> rowIdxs;
+
+    // Grant DataObjectFactory access to the private constructors and
+    // destructors.
+    template<class DataType, typename ... ArgTypes>
+    friend DataType *DataObjectFactory::create(ArgTypes ...);
+
+    template<class DataType>
+    friend void DataObjectFactory::destroy(const DataType *obj);
+
+    /**
+     * @brief Creates a `COOMatrix` and allocates enough memory for the
+     * specified size in the internal `values`, `colIdxs`, and `rowIdxs`
+     * arrays.
+     *
+     * @param numRows The exact number of rows.
+     * @param numCols The exact number of columns.
+     * @param maxNumNonZeros The maximum number of non-zeros in the matrix.
+     * @param zero Whether the allocated memory of the internal arrays shall be
+     * initialized to zeros (`true`), or be left uninitialized (`false`).
+     */
+    COOMatrix(size_t numRows, size_t numCols, size_t maxNumNonZeros, bool zero) :
+            Matrix<ValueType>(numRows, numCols),
+            maxNumNonZeros(maxNumNonZeros),
+            lowerRow(0),
+            upperRow(numRows),
+            view(false),
+            values(new ValueType[maxNumNonZeros + 1], std::default_delete<ValueType[]>()),
+            colIdxs(new size_t[maxNumNonZeros + 1], std::default_delete<size_t[]>()),
+            rowIdxs(new size_t[maxNumNonZeros + 1], std::default_delete<size_t[]>()) {
+        if (zero) {
+            memset(values.get(), 0, (maxNumNonZeros + 1) * sizeof(ValueType));
+            memset(colIdxs.get(), 0, (maxNumNonZeros + 1) * sizeof(size_t));
+            memset(rowIdxs.get(), 0, (maxNumNonZeros + 1) * sizeof(size_t));
+        }
+
+        values.get()[0] = ValueType(0);
+        rowIdxs.get()[0] = size_t(-1);
+        colIdxs.get()[0] = size_t(-1);
```

Comment:
This way of marking the current "head" of the array does not provide direct access to its length. Keeping the length in a separate value that is shared between views would make it more accessable (see review for more detail).
```suggestion
        // attribute before ctor
        std::shared_ptr<size_t> numNonZeros;
        // ctor
        ..., numNonZeros(std::make_shared<size_t>(0)), ...
```